darshan-util/pydarshan/darshan/experimental/plots/data_access_by_filesystem.py

```diff
@@ -0,0 +1,582 @@
+
+    Raises:
+    -------
+    NotImplementedError: for unsupported modules
+
+    ValueError: when POSIX module data is absent from the report
+    """
+
+    if not mod == 'POSIX':
+        raise NotImplementedError("Only the POSIX module is currently supported")
+    if 'POSIX' not in report.modules:
+        raise ValueError("POSIX module data is required")
+
+    # filter the DarshanReport into two "counters" dataframes
+    # with read/write activity in each row (at least 1
+    # byte read or written per row) and extra columns for
+    # filesystem root and filename data
+    df_reads, df_writes = rec_to_rw_counter_dfs_with_cols(report=report,
+                                                          file_id_dict=file_id_dict,
+                                                          mod=mod)
+    read_groups, write_groups = processing_func(df_reads=df_reads,
+                                                df_writes=df_writes)
+    # if either of the Series are effectively empty we want
+    # to produce a new Series with the filesystem_root indices
+    # and count values of 0 (for plotting purposes)
+    read_groups, write_groups = check_empty_series(read_groups=read_groups,
+                                                   write_groups=write_groups,
+                                                   filesystem_roots=filesystem_roots)
+    if verbose:
+        print("read_groups:\n", read_groups)
+        print("write_groups:\n", write_groups)
+    return (read_groups, write_groups)
+
+
+def plot_data(fig: Any,
+              file_rd_series,
+              file_wr_series,
+              bytes_rd_series,
+              bytes_wr_series,
+              filesystem_roots: Sequence[str]):
+    """
+    Produce the horizontal bar plots for the data
+    access by category analysis.
+
+    Parameters
+    ----------
+
+    fig: matplotlib ``figure`` object into which the
+    bar plots will be placed
+
+    file_rd_series: a ``pd.Series`` object with per-category
+    counts of unique files from which at least 1 byte has been
+    read
+
+    file_wr_series: a ``pd.Series`` object with per-category
+    counts of unique files to which at least 1 byte has been
+    written
+
+    bytes_rd_series: a ``pd.Series`` object with per-category
+    counts of bytes read
+
+    bytes_wr_series: a ``pd.Series`` object with per-category
+    counts of bytes written
+
+    filesystem_roots: a sequence of strings containing unique filesystem root paths
+    """
+    list_byte_axes: list = []
+    list_count_axes: list = []
+    # use log10 scale if range exceeds
+    # two orders of magnitude in a column
+    use_log = [False, False]
+    for idx, series_pair in enumerate([[bytes_rd_series, bytes_wr_series],
+                                       [file_rd_series, file_wr_series]]):
+        maxval = max(series_pair[0].max(), series_pair[1].max())
+        minval = min(series_pair[0].min(), series_pair[1].min())
+        ratio = maxval / minval
```

Comment:
tests pass, but careful here--`0/0` is possible I think; NumPy seems graceful enough with just a warning when that kind of thing happens but something to keep in mind..

Reply:
I'll resolve this since I think the symmetric log scaling fixed the above case.